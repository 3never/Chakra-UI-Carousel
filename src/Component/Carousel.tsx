import React from "react";
import {Box} from '@chakra-ui/react'

const Carousel = () => {
  return <Box>
    Carousel
  </Box>
}

export default Carousel