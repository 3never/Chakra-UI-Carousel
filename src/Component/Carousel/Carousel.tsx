import React from "react";
import Item from "../Item";
import { Provider } from "../Provider";

export interface CarouselPropTypes {
  children: React.ReactNode[];
  gap: number;
}

const Carousel: React.FC<CarouselPropTypes> = ({ children, gap }) => {
  return (
    <Provider gap={gap}>
      {children.map((child, index) => (
        <Item gap={gap} key={index} index={index}>
          {child}
        </Item>
      ))}
    </Provider>
  );
};

export default Carousel;
