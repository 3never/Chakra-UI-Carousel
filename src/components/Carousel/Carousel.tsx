import { useMediaQuery, useTheme } from "@chakra-ui/react";
import React, { useContext, useEffect, useMemo } from "react";
import Item from "../Item";
import { Context, ContextType } from "../Provider";
import Track from "../Track";

export interface CarouselPropTypes {
  children: React.ReactNode[];
  gap: number;
}

const Carousel: React.FC<CarouselPropTypes> = ({ children, gap }) => {
  const context = useContext(Context);

  const { setItemWidth, sliderWidth, setMultiplier, setConstraint, itemWidth } =
    context as ContextType;

  const { breakpoints } = useTheme();

  const positions = useMemo(
    () => children?.map((_, index) => -Math.abs((itemWidth + gap) * index)),
    [children, itemWidth, gap]
  );

  const [isBetweenBaseAndMd] = useMediaQuery(
    `(min-width: ${breakpoints?.base}) and (max-width: ${breakpoints?.md})`
  );

  const [isBetweenMdAndXl] = useMediaQuery(
    `(min-width: ${breakpoints?.md}) and (max-width: ${breakpoints?.xl})`
  );

  const [isGreaterThanXL] = useMediaQuery(`(min-width: ${breakpoints?.xl})`);

  useEffect(() => {
    if (isBetweenBaseAndMd) {
      setItemWidth(sliderWidth - gap);
      setMultiplier(0.65);
      setConstraint(1);
    }
    if (isBetweenMdAndXl) {
      setItemWidth(sliderWidth / 2 - gap);
      setMultiplier(0.5);
      setConstraint(2);
    }
    if (isGreaterThanXL) {
      setItemWidth(sliderWidth / 3 - gap);
      setMultiplier(0.35);
      setConstraint(3);
    }
  }, [
    isBetweenBaseAndMd,
    isBetweenMdAndXl,
    isGreaterThanXL,
    sliderWidth,
    gap,
    setItemWidth,
    setMultiplier,
    setConstraint,
  ]);
  return (
    <Track positions={positions}>
      {children.map((child, index) => (
        <Item positions={positions} gap={gap} key={index} index={index}>
          {child}
        </Item>
      ))}
    </Track>
  );
};

export default Carousel;
