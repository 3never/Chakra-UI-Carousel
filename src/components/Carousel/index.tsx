import { useMediaQuery, useTheme } from "@chakra-ui/react";
import React, { useContext, useEffect } from "react";
import Item from "../Item";
import { Context, ContextType } from "../Provider";
import Slider from "../Slider";
import Track from "../Track";

export interface CarouselPropTypes {
  children: React.ReactNode[];
  gap: number;
  full: boolean;
}

export const Carousel: React.FC<CarouselPropTypes> = ({ children, gap, full }) => {
  const context = useContext(Context);

  const {
    setItemWidth,
    sliderWidth,
    setMultiplier,
    setConstraint,
    itemWidth,

    activeItem,
    setPositions,
  } = context as ContextType;

  const { breakpoints } = useTheme();

  useEffect(() => {
    const newPositions = children?.map(
      (_, index) => -Math.abs((itemWidth + gap) * index)
    );
    setPositions(newPositions);
  }, [children, gap, itemWidth, setPositions]);

  const [isBetweenBaseAndMd] = useMediaQuery(
    `(min-width: ${breakpoints?.base}) and (max-width: ${breakpoints?.md})`
  );

  const [isBetweenMdAndXl] = useMediaQuery(
    `(min-width: ${breakpoints?.md}) and (max-width: ${breakpoints?.xl})`
  );

  const [isGreaterThanXL] = useMediaQuery(`(min-width: ${breakpoints?.xl})`);

  useEffect(() => {
    if (full) {
      setItemWidth(sliderWidth - gap);
      setMultiplier(0.8);
      setConstraint(1);
    }else{
      if (isBetweenBaseAndMd) {
        setItemWidth(sliderWidth - gap);
        setMultiplier(0.65);
        setConstraint(1);
      }
      if (isBetweenMdAndXl) {
        setItemWidth(sliderWidth / 2 - gap);
        setMultiplier(0.5);
        setConstraint(2);
      }
      if (isGreaterThanXL) {
        setItemWidth(sliderWidth / 3 - gap);
        setMultiplier(0.35);
        setConstraint(3);
      }
    }
    
  }, [
    isBetweenBaseAndMd,
    isBetweenMdAndXl,
    isGreaterThanXL,
    sliderWidth,
    gap,
    setItemWidth,
    setMultiplier,
    setConstraint,
  ]);

  return (
    <Slider gap={gap}>
      <Track>
        {children.map((child, index) => (
          <Item gap={gap} key={index} index={index}>
            {child}
          </Item>
        ))}
      </Track>
    </Slider>
  );
};
