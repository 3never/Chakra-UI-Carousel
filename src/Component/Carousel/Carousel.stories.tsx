import React from "react";
import Carousel from "./Carousel";
import { Story } from "@storybook/react";
import { Box, Text } from "@chakra-ui/react";

export default {
  title: "Carousel",
  component: Carousel,
  parameters: {
    chromatic: {
      disableSnapshot: false,
    },
  },
};

export const DemoComp = ({ index }: { index: number }) => {
  return (
    <Box borderWidth={2} borderColor='black' w={'10rem'}>
      <Text>Component {index}</Text>
    </Box>
  );
};

export const Primary: Story = () => (
  <Carousel gap={3}>
    <DemoComp index={1} />
    <DemoComp index={2} />
    <DemoComp index={3} />
    <DemoComp index={4} />
  </Carousel>
);
