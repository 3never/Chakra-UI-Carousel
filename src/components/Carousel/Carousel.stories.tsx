import React from "react";
import Carousel from "./Carousel";
import { Story } from "@storybook/react";
import { Box, Text } from "@chakra-ui/react";
import { Provider } from "../Provider";

export default {
  title: "Carousel",
  component: Carousel,
  parameters: {
    chromatic: {
      disableSnapshot: false,
    },
  },
};

export const DemoComp = ({ index }: { index: number }) => {
  return (
    <Box borderWidth={2} borderColor="black" w={"10rem"}>
      <Text>Component {index}</Text>
    </Box>
  );
};

export const Primary: Story = () => (
  <Provider>
    <Carousel gap={3}>
      <DemoComp index={1} />
      <DemoComp index={2} />
      <DemoComp index={3} />
      <DemoComp index={4} />
    </Carousel>
  </Provider>
);
