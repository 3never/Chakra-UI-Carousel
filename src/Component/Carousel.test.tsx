import React from 'react';
import Carousel from './Carousel';
import renderer from 'react-test-renderer';


describe('renders correctly', () => {
  it('Carousel renders correctly', () => {
    const container = renderer.create(<Carousel />).toJSON();
    expect(container).toMatchSnapshot();
  });
});
