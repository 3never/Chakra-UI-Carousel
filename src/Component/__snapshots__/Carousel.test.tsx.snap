// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Carousel renders correctly 1`] = `
<div
  className="css-13d2904"
>
  <p
    className="chakra-text css-x2jib2"
  >
    Carousel
  </p>
</div>
`;
